### ISSUE TITLE
Problem running on a router

### ISSUE THREAD

**Issue by Migel1** [artem-russkikh/wireproxy-awg#5]
I tried to run wireproxy on a router.
To achieve this I crosscompiled source code for ARM Linux (GOOS=linux GOARCH=arm GOARM=5) and run the resulting WireProxy binary on the router.

The binary started successfully, but when I tried to connect to the proxy I got the following error on the router console:
```
DEBUG: 2024/12/16 17:51:42 peer(bmXO…fgyo) - Sending handshake initiation
ERROR: 2024/12/16 17:51:42 peer(bmXO…fgyo) - Failed to send handshake initiation: write udp 0.0.0.0:61895: sendmmsg: function not implemented
```

The kernel version on router is `Linux router 2.6.36.4brcmarm #82 PREEMPT Wed Aug 7 09:46:26 CEST 2024 armv7l Tomato`.

I this Linux kernel supported? Or may be I need to add some special options to run wireproxy on a router?

**Comment by artem-russkikh** [comment]
I didn't attempt to run wireproxy on a router. Need to research this topic. Meanwhile you can search for similar issues in the origin wireguard GitHub repository at https://github.com/pufferffish/wireproxy

**Comment by Migel1** [comment]
It seems that the gvisor library uses the sendmmsg Linux system call, which is available only in Linux kernels 3.0+ :(

**Comment by ValentineShilov** [comment]
In Amnezia-wg I fixed the same problem by changing conn/bind_std.go.
In  func (s *StdNetBind) send(conn *net.UDPConn, pc batchWriter, msgs []ipv6.Message)
by default for Linux WriteBatch is used to send udp packets (and it uses syscall absent on kernel 2.*)
change the code `
var (
		n     int
		err   error
		start int
	)
if runtime.GOOS == "linux" || runtime.GOOS == "android" {
		for {
			n, err = pc.WriteBatch(msgs[start:], 0)
			if err != nil || n == len(msgs[start:]) {
				break
			}
			start += n
		}
	} else {
		for _, msg := range msgs {
			_, _, err = conn.WriteMsgUDP(msg.Buffers[0], msg.OOB, msg.Addr.(*net.UDPAddr))
			if err != nil {
				break
			}
		}
	}`
to 
`
var (
err   error
)
      for _, msg := range msgs {
	      _, _, err = conn.WriteMsgUDP(msg.Buffers[0], msg.OOB, msg.Addr.(*net.UDPAddr))
	      if err != nil {
		      break
	      }
		
	}`

**Comment by ValentineShilov** [comment]
> I tried to run wireproxy on a router. To achieve this I crosscompiled source code for ARM Linux (GOOS=linux GOARCH=arm GOARM=5) and run the resulting WireProxy binary on the router.
> 

Made build compatible with old kernels for my fork, you can try if it works with your router https://github.com/ValentineShilov/wireproxy-awg-udp/releases    (wireproxy-arm5-kernel2.6 )

**Comment by ValentineShilov** [comment]
> без этих манипуляций бинарник для arm64 ?

Добавил
PS спасибо, что обратили внимание, открыл Issues

**Comment by Migel1** [comment]
> Made build compatible with old kernels for my fork, you can try if it works with your router https://github.com/ValentineShilov/wireproxy-awg-udp/releases (wireproxy-arm5-kernel2.6 )

Yes, this version works without problems.

**Comment by Daariush** [comment]
Will it run on a openwrt?

**Comment by skaunov** [comment]
Yes, kudos to the author(s). Someone who knows how to make it (more) visible on the original/upstream resources will do a good deed. Also cross-compilation and Open-WRT service can be added as future issues (if someone is interested).

**Comment by anton-vinogradov** [comment]
Thanks for the detailed report and for the thread — this is enough to pin it down precisely.

Root cause: the failing `sendmmsg` call isn't in wireproxy itself. We build the device with `conn.NewDefaultBind()`, which comes from the dependency `github.com/amnezia-vpn/amneziawg-go`. Its UDP bind (`conn/bind_std.go`) unconditionally uses batched syscalls on Linux — `sendmmsg(2)` on send (`StdNetBind.send` → `WriteBatch`) and `recvmmsg(2)` on receive (`StdNetBind.receiveIP` → `ReadBatch`). Both require kernel ≥ 3.0, so on your 2.6.36 Tomato kernel they fail with `function not implemented`. There's no kernel-version fallback in that code, so it can't be worked around from wireproxy's side without overriding the dependency.

What works today: @ValentineShilov's `wireproxy-arm5-kernel2.6` build (you already confirmed it runs). The change is exactly the one he posted — replace the batched `WriteBatch`/`ReadBatch` with per-message `conn.WriteMsgUDP`/`conn.ReadMsgUDP`. Functionally identical on a one-packet-at-a-time basis; you only give up the batching micro-optimization, which old kernels can't do anyway.

Proper fix path: since the code lives in amneziawg-go, the clean fix is to land a `WriteBatch→WriteMsgUDP` / `ReadBatch→ReadMsgUDP` fallback there — ideally guarded so modern kernels keep the batched fast path (e.g. a build tag, or a one-time `ENOSYS` detection) — and then bump the dependency here. I confirmed locally that this change cross-compiles for your exact target (`GOOS=linux GOARCH=arm GOARM=5`) and produces a working binary. Until it lands upstream, the kernel2.6 community build above is the practical option for old routers / OpenWrt.
